### ISSUE TITLE
Confusion with DOM -> Actions -> Reducers (on my side?)

### ISSUE THREAD

**Issue by baranovxyz** [SteveALee/cycle-spa-template#1]
First of all, really BIG THANK YOU for a good template :)

I've spent some time building a prototype using it and other sources (like github.com/staltz/matrixmultiplication.xyz which is said to be a 'a good learning resource on how to structure Cycle.js apps').

And I have had a LOT of confusion.
Mostly with the fact, that in your template intent function returns Stream of Reducer, while according to official docs (whttps://cycle.js.org/model-view-intent.html) and examples it should return Stream of Action which is then feeded into model() that returns Stream of Reducer.

So here is a question - if that was intentional, than - why you decided to do that way?
And if not, I can refactor and make a pull request so that it aligns with the docs...

**Comment by SteveALee** [comment]
Hi @baranovxyz thanks for your kind words! I'll try to clarify.

1st though, note that the latest version of the [one-fits-all flavor](https://github.com/cyclejs-community/one-fits-all/tree/spa-continue) for create-cycle-app is being modified to include some of the ideas here. @jvanbruegge is doing a fantastic job and I think you will find that even more useful as a learning tool.

So your confusion is understandable as there are 2 distinct ways of handling state in cyclejs Apps. 
1) inline in a component using a model() function and often [xstream fold](https://github.com/staltz/xstream#fold) operator
2) global single state atom using [onionify](https://github.com/staltz/cycle-onionify) (this is like redux if you are familiar with that)

With the early examples and talks @staltz emphasized the 1st method as being more 'reactive' and having no need to store state outside the streams. Subsequently, the community have been experimenting with 2) and it is likely to become the more standard approach.

While both approaches fit the MVI concept (events -> actions -> state -> view), you see a model function function in 1) you don't in when onionify is used.

If you use onionify you just need to [wrap your top level component](https://cycle.js.org/model-view-intent.html) and each component then magically gets a 'onion' source and sink. A component then gets the latest state via the [onion stream](https://github.com/SteveALee/cycle-spa-template/blob/master/src/pages/page2.tsx#L16) which it passes to the view. It also returns state updates (derived from its actions) to the onion as a [stream of reducer functions](https://github.com/SteveALee/cycle-spa-template/blob/master/src/pages/page2.tsx#L32), as you spotted. A little magic then makes sure a component does not usually get to see state for components above it in the hierarchy (but there are options to modify this).

So to recap, both state handling methods are MVI but in 1) the Model is inline as a function in your component and in 2) it is external to your component, in the onion.

Finally it is common to see the 2nd form done in a slightly different way by splitting out actions explicit messages (objects) and action handler functions (again, influenced by redux). In this scheme, all the actions are collected in to one stream internal to the component and each handler filters those actions it is interested in. See the new version of the [speaker component](https://github.com/cyclejs-community/one-fits-all/blob/spa-continue/template/src/components/speaker.tsx) I linked to above. 
I hope that helps and I'll perhaps make this a blog post:)

Do hop on to gitter if you have questions.

**Comment by baranovxyz** [comment]
Yeah, it make sense... 

I added some component and onionified them, and didnt add typings to them, thinking that ts is smart enough, now I know it is not :)

I think the main issue I had is that in your example function intent() was returning a stream of Reducers while in other examples (from where I also borrowed code it was returning stream of actions) and I xs.merge both and so I have spent time digging it down :)

Just didnt want that happen for another person.

**Comment by jvanbruegge** [comment]
Thats one of the biggest reasons to use Typescript. You as a programmer will make dumb mistakes, a type system helps a lot

**Comment by jvanbruegge** [comment]
I usually let my intent return either a strwam of objects or a reducer depending on what I want to do with them. If I only want to use them with onionify, introducing a bijective mapping from object to reducer is boilerplate.
If I have a lot of different stuff (http, oniin, maybe modal, etc) i use objects as those provide a common interface for all

**Comment by SteveALee** [comment]
I guess there are no "hard and fast" rules for how you structure you components. Just conventions that if followed will reduce learning curves. :)

**Comment by SteveALee** [comment]
> Thats one of the biggest reasons to use Typescript. You as a programmer will make dumb mistakes, a type system helps a lot

And if you an editor with intellisense like VS Code it really helps improve the editing experience by giving you choices and showing errors as you edit.
